src/sage/schemes/weighted_projective/weighted_projective_space.py

```diff
@@ -0,0 +1,420 @@
+            raise TypeError("the zero vector is not a point in projective space")
+
+        return True
+
+    def coordinate_ring(self) -> PolynomialRing_generic:
+        """
+        Return the coordinate ring of this weighted projective space.
+
+        EXAMPLES::
+
+            sage: WP = WeightedProjectiveSpace(GF(19^2, 'α'), [1, 3, 4, 1], 'abcd')
+            sage: # needs sage.rings.finite_rings
+            sage: R = WP.coordinate_ring(); R
+            Multivariate Polynomial Ring in a, b, c, d over Finite Field in α of size 19^2
+            sage: R.term_order()
+            Weighted degree reverse lexicographic term order with weights (1, 3, 4, 1)
+
+        ::
+
+            sage: WP = WeightedProjectiveSpace(QQ, [1, 1, 1], ['alpha', 'beta', 'gamma'])
+            sage: R = WP.coordinate_ring(); R
+            Multivariate Polynomial Ring in alpha, beta, gamma over Rational Field
+            sage: R.term_order()
+            Weighted degree reverse lexicographic term order with weights (1, 1, 1)
+        """
+        if not hasattr(self, "_coordinate_ring"):
+            term_order = TermOrder("wdegrevlex", self.weights())
+            self._coordinate_ring = PolynomialRing(
+                self.base_ring(),
+                self.dimension_relative() + 1,
+                names=self.variable_names(),
+                order=term_order,
+            )
+        return self._coordinate_ring
+
+    def _validate(self, polynomials):
+        """
+        If ``polynomials`` is a tuple of valid polynomial functions on ``self``,
+        return ``polynomials``, otherwise raise ``TypeError``.
+
+        Since this is a weighted projective space, polynomials must be
+        homogeneous with respect to the grading of this space.
+
+        INPUT:
+
+        - ``polynomials`` -- tuple of polynomials in the coordinate ring of
+            this space.
+
+        OUTPUT:
+
+        - tuple of polynomials in the coordinate ring of this space.
+
+        EXAMPLES::
+
+            sage: P.<x, y, z> = WeightedProjectiveSpace(QQ, [1, 3, 1])
+            sage: P._validate([x*y - z^4, x])
+            [x*y - z^4, x]
+            sage: P._validate([x*y - z^2, x])
+            Traceback (most recent call last):
+            ...
+            TypeError: x*y - z^2 is not homogeneous with weights (1, 3, 1)
+            sage: P._validate(x*y - z)
+            Traceback (most recent call last):
+            ...
+            TypeError: the argument polynomials=x*y - z must be a list or tuple
+        """
+        if not isinstance(polynomials, (list, tuple)):
+            raise TypeError(
+                f"the argument polynomials={polynomials} must be a list or tuple"
+            )
+
+        R = self.coordinate_ring()
+        for f in map(R, polynomials):
+            if not f.is_homogeneous():
+                raise TypeError(f"{f} is not homogeneous with weights {self.weights()}")
+
+        return polynomials
+
+    def _latex_(self):
```

Comment:
For function where it's very clear what the returned type is, why not add a return type annotation? You can do this for all instances of _repr_ and _latex_ in this PR (I noticed you already have it for some).

```suggestion
    def _latex_(self) -> str:
```